File: server/engine/templates/template_moco.py
import opensim as osim
import matplotlib
matplotlib.use('Agg')

# Update the model.
# -----------------
# Replace locked coordinates with weld joints.
model = osim.Model('../Models/final.osim')
model.initSystem()
coordinates = model.getCoordinateSet()
locked_joints = list()
locked_coordinates = list()
for icoord in range(coordinates.getSize()):
    coord = coordinates.get(icoord)
    if coord.get_locked():
        locked_coordinates.append(coord.getName())
        locked_joints.append(coord.getJoint().getName())

# Remove actuators associated with locked coordinates.
for coordinate in locked_coordinates:
    forceSet = model.updForceSet()
    for iforce in range(forceSet.getSize()):
        force = forceSet.get(iforce)
        if force.getConcreteClassName().endswith('CoordinateActuator'):
            actu = osim.CoordinateActuator.safeDownCast(force)
            if actu.get_coordinate() == coordinate:
                forceSet.remove(iforce)
                break

model.finalizeConnections()
model.initSystem()

# Construct a ModelProcessor. The default muscles in the model are replaced with
# optimization-friendly DeGrooteFregly2016Muscles, and adjustments are made to the
# default muscle parameters. We also add reserve actuators to the model and apply
# the external loads.
modelProcessor = osim.ModelProcessor(model)
modelProcessor.append(osim.ModOpReplaceJointsWithWelds(locked_joints))
modelProcessor.append(osim.ModOpAddExternalLoads('../ID/@TRIAL@_external_forces.xml'))
modelProcessor.append(osim.ModOpIgnoreTendonCompliance())
modelProcessor.append(osim.ModOpReplaceMusclesWithDeGrooteFregly2016())
modelProcessor.append(osim.ModOpIgnorePassiveFiberForcesDGF())
modelProcessor.append(osim.ModOpAddReserves(50.0))

# Construct the MocoInverse tool.
# -------------------------------
inverse = osim.MocoInverse()

# Set the model processor and time bounds.
inverse.setModel(modelProcessor)
inverse.set_initial_time(@INITIAL_TIME@)
inverse.set_final_time(@FINAL_TIME@)

# Load the kinematics data source.
# Add in the Rajagopal2015 patella coordinates to the kinematics table, if missing.
table = osim.TimeSeriesTable('../IK/@TRIAL@_ik.mot')
labels = table.getColumnLabels()
if 'knee_angle_r' in labels and 'knee_angle_r_beta' not in labels:
    table.appendColumn('knee_angle_r_beta', table.getDependentColumn('knee_angle_r'))
if 'knee_angle_l' in labels and 'knee_angle_l_beta' not in labels:
    table.appendColumn('knee_angle_l_beta', table.getDependentColumn('knee_angle_l'))

# Construct a TableProcessor to update the kinematics table labels to use absolute path names.
# Add the kinematics to the MocoInverse tool.
tableProcessor = osim.TableProcessor(table)
tableProcessor.append(osim.TabOpUseAbsoluteStateNames())
inverse.setKinematics(tableProcessor)

# Configure additional settings for the MocoInverse problem including the mesh
# interval, convergence tolerance, constraint tolerance, and max number of iterations.
inverse.set_mesh_interval(0.02)
inverse.set_convergence_tolerance(1e-5)
inverse.set_constraint_tolerance(1e-5)
inverse.set_max_iterations(2000)
# Skip any extra columns in the kinematics data source.
inverse.set_kinematics_allow_extra_columns(True)

# Solve the problem.
# ------------------
# Solve the problem and write the solution to a Storage file.
solution = inverse.solve()
solution.getMocoSolution().write('@TRIAL@_moco.sto')

# Generate a PDF with plots for the solution trajectory.
model = modelProcessor.process()
report = osim.report.Report(model,
                            '@TRIAL@_moco.sto',
                            output='@TRIAL@_moco.pdf',
                            bilateral=True)
# The PDF is saved to the working directory.
report.generate()

File: server/engine/templates/template_moco.m
import org.opensim.modeling.*;

% Update the model.
% -----------------
% Replace locked coordinates with weld joints.
model = Model('../Models/final.osim');
model.initSystem();
coordinates = model.getCoordinateSet();
locked_joints = StdVectorString();
locked_coordinates = {};
for icoord = 0:coordinates.getSize()-1
    coord = coordinates.get(icoord);
    if coord.get_locked()
        locked_coordinates{end+1} = char(coord.getName());
        locked_joints.add(coord.getJoint().getName());
    end
end

% Remove actuators associated with locked coordinates.
for i = 1:length(locked_coordinates)
    coordinate = locked_coordinates{i};
    forceSet = model.updForceSet();
    for iforce = 0:forceSet.getSize()-1
        force = forceSet.get(iforce);
        actu = CoordinateActuator.safeDownCast(force);
        if ~isempty(actu) 
            if strcmp(actu.get_coordinate(), coordinate)
                forceSet.remove(iforce);
                break;
            end
        end
    end
end

model.finalizeConnections();
model.initSystem();

% Construct a ModelProcessor.
modelProcessor = ModelProcessor(model);
modelProcessor.append(ModOpReplaceJointsWithWelds(locked_joints));
modelProcessor.append(ModOpAddExternalLoads('../ID/@TRIAL@_external_forces.xml'));
modelProcessor.append(ModOpIgnoreTendonCompliance());
modelProcessor.append(ModOpReplaceMusclesWithDeGrooteFregly2016());
modelProcessor.append(ModOpIgnorePassiveFiberForcesDGF());
modelProcessor.append(ModOpAddReserves(50.0));

% Construct the MocoInverse tool.
inverse = MocoInverse();

% Set the model processor and time bounds.
inverse.setModel(modelProcessor);
inverse.set_initial_time(@INITIAL_TIME@);
inverse.set_final_time(@FINAL_TIME@);

% Load the kinematics data source.
table = TimeSeriesTable('../IK/@TRIAL@_ik.mot');
labels = table.getColumnLabels();
for i = 0:labels.size()-1
    if strcmp(char(labels.get(i)), 'knee_angle_r')
        table.appendColumn('knee_angle_r_beta', ...
            table.getDependentColumn('knee_angle_r'));
    end
    if strcmp(char(labels.get(i)), 'knee_angle_l')
        table.appendColumn('knee_angle_l_beta', ...
            table.getDependentColumn('knee_angle_l'));
    end
end

% Construct a TableProcessor to update the kinematics table labels.
tableProcessor = TableProcessor(table);
tableProcessor.append(TabOpUseAbsoluteStateNames());
inverse.setKinematics(tableProcessor);

% Configure additional settings for the MocoInverse problem.
inverse.set_mesh_interval(0.02);
inverse.set_convergence_tolerance(1e-5);
inverse.set_constraint_tolerance(1e-5);
inverse.set_max_iterations(2000);
inverse.set_kinematics_allow_extra_columns(true);

% Solve the problem.
% ------------------
% Solve the problem and write the solution to a Storage file.
solution = inverse.solve();
solution.getMocoSolution().write('@TRIAL@_moco.sto');